### ISSUE TITLE
Inner tool attributes emit spurious unused attribute warning

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#81661]
I tried this code:

```rust
// main.rs
mod bar;

fn main() {
    bar::foo();
}
```

```rust
// bar.rs
#![rustfmt::skip]

pub fn foo() { println!("unformatted") }
```

I expected to see this happen: No warnings and `bar.rs` will not get formatted by `cargo fmt`.

Instead, this happened: `cargo build` emits a warning, `bar.rs` is correctly not formatted by `cargo fmt`.

```
warning: unused attribute
 --> src/bar.rs:1:1
  |
1 | #![rustfmt::skip]
  | ^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default
```

While trying to minimize I changed `mod bar` to be an inline module, but that hits the unstable `custom_inner_attributes` feature (https://github.com/rust-lang/rust/issues/54726). So maybe `rustfmt` _shouldn't_ be obeying this attribute? Either way there's an inconsistency here with whether `rustc` or `rustfmt` believe the attribute is active or not.

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (0f6f2d681 2020-12-06)
binary: rustc
commit-hash: 0f6f2d681b39c5f95459cd09cb936b6ceb27cd82
commit-date: 2020-12-06
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

**Comment by ehuss** [comment]
I believe that in other positions, tool attributes like `rustfmt` are only marked as "used" as a side-effect of the macro expansion machinery.  From what I can tell, it looks like inner attributes on file-based modules simply aren't treated as candidates for macro expansion (presumably since that is not supported, so there is no reason to do that).

I imagine @petrochenkov might have a suggestion how how that should be handled.  I'm uncertain if it makes sense to somehow just mark all tool attributes as used from within the resolver, or if the unused attributes lint should ignore tool attributes (I'm guessing the latter is harder because only the resolver knows the set of tool names AFAIK).

More details:
Attribute expansion normally hits [this code block](https://github.com/rust-lang/rust/blob/368275062fb655c1f36e0398f88b15379a1f3c93/compiler/rustc_expand/src/expand.rs#L1361-L1365) for things like `#[rustfmt::skip] mod inline {}`, but since file-based modules are handled later in that function [here](https://github.com/rust-lang/rust/blob/368275062fb655c1f36e0398f88b15379a1f3c93/compiler/rustc_expand/src/expand.rs#L1395-L1416), the module attributes are not treated as macro invocation candidates.

**Comment by petrochenkov** [comment]
Inner attributes on file-based modules should certainly participate in expansion.

Could someone check whether this is a regression from https://github.com/rust-lang/rust/pull/69838 or a pre-existing issue?

**Comment by Nemo157** [comment]
Before 1.44 this hits `error[E0658]: non-builtin inner attributes are unstable`, adding `#![feature(custom_inner_attributes)]` on `rustc 1.43.0-nightly (823ff8cf1 2020-03-07)` gives no warning.

**Comment by petrochenkov** [comment]
The regression seems pretty bad because it creates a hole in the `#![feature(custom_inner_attributes)]` feature gate, so unstable inner attributes are now accepted.

**Comment by Nemo157** [comment]
Checking some more dates; `rustc 1.44.0-nightly (f4c675c47 2020-03-19)` both added the warning and removed the error, so it seems likely that #69838 accidentally stabilized inner attributes (unless there was another PR touching that code in that exact nightly).

**Comment by ehuss** [comment]
Ah.  So just to be clear, here's the behavior that I see:

## Inner attribute on file-based module

Proc-macro attribute:
1.43: Requires `feature(custom_inner_attributes, proc_macro_hygiene)`, works as expected
1.44: Does not require a feature, the proc-macro is ignored, and issues an unused attribute warning.
1.51: Same as 1.44

rustfmt::skip:
1.43: Requires `feature(custom_inner_attributes)`, no warnings with the feature
1.44: Does not require a feature, but issues an unused attribute warning
1.51: Same as 1.44

## Inner attributes on inline modules

Proc-macro attribute:
1.43: Requires `feature(custom_inner_attributes)`, works as expected
1.44: Same as 1.43
1.51: Same as 1.43

rustfmt::skip:
1.43: Requires `feature(custom_inner_attributes)`, no warnings with the feature
1.44: Same as 1.43
1.51: Same as 1.43

---

It is not immediately clear to me how #69838 changed that (obviously it is touching the relevant code, but I'm uncertain how it worked prior to that, and I don't have time right now to investigate more).

**Comment by apiraino** [comment]
Assigning `P-high` as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2381661.20Inner.20tool.20attributes.20emit.20spurious.20unused.20attribute.E2.80.A6) and removing `I-prioritize`.

**Comment by petrochenkov** [comment]
Fixed in https://github.com/rust-lang/rust/pull/82399.
